File: src/print_partner/core/merge.py
"""Layered merge engine for build profiles."""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

import trimesh
from rapidfuzz import fuzz

from print_partner.core.scanner import ScannedPart, normalize_match_key


@dataclass
class MergePart:
    match_key: str
    relative_path: str
    filename: str
    source_layer: str
    status: str  # base|added|replaced|excluded|conflict
    role: str
    quantity_auto: int
    part_slug: str
    included: bool = True
    quantity_override: int | None = None
    notes: str = ""
    geometry_same: bool | None = None
    absolute_path: Path | None = None
    filament_color_id: str | None = None
    filament_display: str = ""
    filament_hex: str | None = None
    filament_swatch_url: str = ""

    @property
    def quantity_effective(self) -> int:
        if self.quantity_override is not None:
            return self.quantity_override
        return self.quantity_auto


@dataclass
class MergeResult:
    parts: list[MergePart]
    duplicate_hints: list[tuple[str, str, float]] = field(default_factory=list)


def _scanned_to_merge(part: ScannedPart, status: str, source: str) -> MergePart:
    return MergePart(
        match_key=part.match_key,
        relative_path=part.relative_path,
        filename=part.filename,
        source_layer=source,
        status=status,
        role=part.role,
        quantity_auto=part.quantity,
        part_slug=part.part_slug,
        absolute_path=part.absolute_path,
    )


def compare_geometry(path_a: Path | None, path_b: Path | None) -> bool | None:
    if path_a is None or path_b is None:
        return None
    if not path_a.is_file() or not path_b.is_file():
        return None
    try:
        mesh_a = trimesh.load(path_a, force="mesh")
        mesh_b = trimesh.load(path_b, force="mesh")
        if not isinstance(mesh_a, trimesh.Trimesh) or not isinstance(mesh_b, trimesh.Trimesh):
            return None
        return mesh_a.vertices.shape == mesh_b.vertices.shape and (
            (mesh_a.vertices - mesh_b.vertices).max() < 1e-4
        )
    except Exception:
        return None


def _find_duplicate_hints(parts: list[MergePart], threshold: float = 85.0) -> list[tuple[str, str, float]]:
    hints: list[tuple[str, str, float]] = []
    slugs = [(p.part_slug, p.match_key) for p in parts if p.included]
    seen: set[tuple[str, str]] = set()
    for i, (slug_a, key_a) in enumerate(slugs):
        for slug_b, key_b in slugs[i + 1 :]:
            if slug_a == slug_b and key_a != key_b:
                continue
            score = fuzz.ratio(slug_a.lower(), slug_b.lower())
            if score >= threshold:
                pair = tuple(sorted((key_a, key_b)))
                if pair not in seen:
                    seen.add(pair)  # type: ignore[arg-type]
                    hints.append((key_a, key_b, score))
    return hints


def merge_layers(
    layer_scans: list[tuple[str, list[ScannedPart]]],
    existing: dict[str, MergePart] | None = None,
) -> MergeResult:
    """
    Apply layers in order: base seeds, addon adds/replaces.
    existing: prior parts keyed by match_key to preserve overrides.
    """
    prior = existing or {}
    merged: dict[str, MergePart] = {}
    slug_index: dict[str, str] = {}

    for layer_idx, (layer_name, scanned) in enumerate(layer_scans):
        is_base = layer_idx == 0
        for part in scanned:
            key = normalize_match_key(part.relative_path)
            prev = merged.get(key)
            old_prior = prior.get(key)

            if prev is None and not is_base:
                status = "added"
            elif prev is None:
                status = "base"
            else:
                status = "replaced"

            mp = _scanned_to_merge(part, status, layer_name)
            if old_prior:
                mp.quantity_override = old_prior.quantity_override
                mp.notes = old_prior.notes
                mp.included = old_prior.included
                if not old_prior.included:
                    mp.status = "excluded"

            if prev and prev.absolute_path and part.absolute_path:
                mp.geometry_same = compare_geometry(prev.absolute_path, part.absolute_path)

            merged[key] = mp

            other_key = slug_index.get(part.part_slug)
            if other_key and other_key != key:
                merged[key].status = "conflict"
                if other_key in merged:
                    merged[other_key].status = "conflict"
            else:
                slug_index[part.part_slug] = key

    parts = list(merged.values())
    hints = _find_duplicate_hints(parts)
    return MergeResult(parts=parts, duplicate_hints=hints)

File: web/packages/domain/src/merge.ts
/** Layered merge engine for build profiles (ported from Python merge.py). */

import { parsePartSlug } from "./parsers.js";
import { normalizeMatchKey, type ScannedPart } from "./scanner.js";
import type { NamingProfile } from "./stl-naming.js";

export type MergePart = {
  matchKey: string;
  relativePath: string;
  filename: string;
  sourceLayer: string;
  status: string;
  role: string;
  quantityAuto: number;
  partSlug: string;
  included: boolean;
  quantityOverride: number | null;
  notes: string;
  geometrySame: boolean | null;
  absolutePath: string | null;
};

export type MergeResult = {
  parts: MergePart[];
  duplicateHints: Array<[string, string, number]>;
};

export function quantityEffective(part: MergePart): number {
  return part.quantityOverride ?? part.quantityAuto;
}

export type SlugConflictPart = {
  matchKey: string;
  relativePath: string;
  filename: string;
  included: boolean;
  partSlug?: string;
};

/** Included parts that share a slug but not the same match key (active merge conflicts). */
export function findActiveSlugConflictKeys(
  parts: SlugConflictPart[],
  profile?: NamingProfile | null,
): Set<string> {
  const slugIndex = new Map<string, string>();
  const conflictKeys = new Set<string>();

  for (const part of parts) {
    if (!part.included) continue;
    const slug =
      part.partSlug?.trim() ||
      parsePartSlug(part.relativePath || part.filename, profile);
    const key = part.matchKey;
    const otherKey = slugIndex.get(slug);
    if (otherKey && otherKey !== key) {
      conflictKeys.add(key);
      conflictKeys.add(otherKey);
    } else if (!otherKey) {
      slugIndex.set(slug, key);
    }
  }

  return conflictKeys;
}

function scannedToMerge(part: ScannedPart, status: string, source: string): MergePart {
  return {
    matchKey: part.matchKey,
    relativePath: part.relativePath,
    filename: part.filename,
    sourceLayer: source,
    status,
    role: part.role,
    quantityAuto: part.quantity,
    partSlug: part.partSlug,
    absolutePath: part.absolutePath,
    included: true,
    quantityOverride: null,
    notes: "",
    geometrySame: null,
  };
}

function levenshtein(a: string, b: string): number {
  const m = a.length;
  const n = b.length;
  const dp: number[][] = Array.from({ length: m + 1 }, () => Array(n + 1).fill(0));
  for (let i = 0; i <= m; i++) dp[i][0] = i;
  for (let j = 0; j <= n; j++) dp[0][j] = j;
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      dp[i][j] = Math.min(dp[i - 1][j] + 1, dp[i][j - 1] + 1, dp[i - 1][j - 1] + cost);
    }
  }
  return dp[m][n];
}

function fuzzRatio(a: string, b: string): number {
  const s1 = a.toLowerCase();
  const s2 = b.toLowerCase();
  const longer = s1.length >= s2.length ? s1 : s2;
  const shorter = s1.length >= s2.length ? s2 : s1;
  if (longer.length === 0) return 100;
  const dist = levenshtein(longer, shorter);
  return Math.round(((longer.length - dist) / longer.length) * 100);
}

function findDuplicateHints(parts: MergePart[], threshold = 85): Array<[string, string, number]> {
  const hints: Array<[string, string, number]> = [];
  const slugs = parts.filter((p) => p.included).map((p) => [p.partSlug, p.matchKey] as const);
  const seen = new Set<string>();
  for (let i = 0; i < slugs.length; i++) {
    const [slugA, keyA] = slugs[i];
    for (let j = i + 1; j < slugs.length; j++) {
      const [slugB, keyB] = slugs[j];
      if (slugA === slugB && keyA !== keyB) continue;
      const score = fuzzRatio(slugA, slugB);
      if (score >= threshold) {
        const pair = [keyA, keyB].sort().join("|");
        if (!seen.has(pair)) {
          seen.add(pair);
          hints.push([keyA, keyB, score]);
        }
      }
    }
  }
  return hints;
}

export function mergeLayers(
  layerScans: Array<[string, ScannedPart[]]>,
  existing: Record<string, MergePart> | null = null,
  options?: { geometryCompare?: boolean },
): MergeResult {
  void options?.geometryCompare;
  const prior = existing ?? {};
  const merged: Record<string, MergePart> = {};
  const slugIndex: Record<string, string> = {};

  for (let layerIdx = 0; layerIdx < layerScans.length; layerIdx++) {
    const [layerName, scanned] = layerScans[layerIdx];
    const isBase = layerIdx === 0;
    for (const part of scanned) {
      const key = normalizeMatchKey(part.relativePath);
      const prev = merged[key];
      const oldPrior = prior[key];

      let status: string;
      if (prev == null && !isBase) status = "added";
      else if (prev == null) status = "base";
      else status = "replaced";

      const mp = scannedToMerge(part, status, layerName);
      if (oldPrior) {
        mp.quantityOverride = oldPrior.quantityOverride;
        mp.notes = oldPrior.notes;
        mp.included = oldPrior.included;
        if (!oldPrior.included) mp.status = "excluded";
      }

      merged[key] = mp;

      const otherKey = slugIndex[part.partSlug];
      if (otherKey && otherKey !== key) {
        merged[key].status = "conflict";
        if (merged[otherKey]) merged[otherKey].status = "conflict";
      } else {
        slugIndex[part.partSlug] = key;
      }
    }
  }

  const parts = Object.values(merged);
  return { parts, duplicateHints: findDuplicateHints(parts) };
}

export class MergeWouldWipeProfileError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MergeWouldWipeProfileError";
  }
}

export type PartProgressRow = {
  id: number;
  unitIndex: number;
  completed: boolean;
};

export function saveMergeResultPlan(
  existingParts: MergePart[],
  result: MergeResult,
): MergePart[] {
  if (!result.parts.length && existingParts.length) {
    throw new MergeWouldWipeProfileError(
      "Scan found no STL files (check Projects → Import files… for each repo). Existing parts were not removed.",
    );
  }
  const existingByKey = new Map(existingParts.map((p) => [p.matchKey, p]));
  const newKeys = new Set(result.parts.map((p) => p.matchKey));
  const out: MergePart[] = [];

  for (const mp of result.parts) {
    const prior = existingByKey.get(mp.matchKey);
    if (prior) {
      mp.quantityOverride = prior.quantityOverride;
      mp.notes = prior.notes || mp.notes;
      mp.included = prior.included;
    }
    out.push(mp);
  }

  void newKeys;
  return out;
}
